modules/AWS/eks/autoscaling.tf

```diff
@@ -0,0 +1,102 @@
+module "autoscaler_iam_role" {
+  source  = "terraform-aws-modules/iam/aws//modules/iam-assumable-role-with-oidc"
+  version = "4.13.2"
+
+  create_role  = true
+  role_name    = "${var.cluster_name}-cluster-autoscaler"
+  provider_url = replace(module.eks.cluster_oidc_issuer_url, "https://", "")
+
+  role_policy_arns = [
+    aws_iam_policy.cluster_autoscaler.arn
+  ]
+  oidc_fully_qualified_subjects = [
+    "system:serviceaccount:${local.autoscaler_service_account_namespace}:${local.autoscaler_service_account_name}"
+  ]
+}
+
+resource "aws_iam_policy" "cluster_autoscaler" {
+  name_prefix = "cluster-autoscaler"
+  description = "EKS cluster-autoscaler policy for cluster ${module.eks.cluster_id}"
+  policy      = data.aws_iam_policy_document.cluster_autoscaler.json
+}
+
+data "aws_iam_policy_document" "cluster_autoscaler" {
+  statement {
+    sid    = "clusterAutoscalerAll"
+    effect = "Allow"
+
+    actions = [
+      "autoscaling:DescribeAutoScalingGroups",
+      "autoscaling:DescribeAutoScalingInstances",
+      "autoscaling:DescribeLaunchConfigurations",
+      "autoscaling:DescribeTags",
+      "ec2:DescribeLaunchTemplateVersions",
+      "ec2:DescribeInstanceTypes"
+    ]
+
+    resources = [
+      "*"
+    ]
+  }
+
+  statement {
+    sid    = "clusterAutoscalerOwn"
+    effect = "Allow"
+
+    actions = [
+      "autoscaling:SetDesiredCapacity",
+      "autoscaling:TerminateInstanceInAutoScalingGroup",
+      "autoscaling:UpdateAutoScalingGroup",
+    ]
+
+    resources = [
+      "*"
+    ]
+
+    condition {
+      test     = "StringEquals"
+      variable = "autoscaling:ResourceTag/kubernetes.io/cluster/${module.eks.cluster_id}"
+      values = [
+        "owned"
+      ]
+    }
+
+    condition {
+      test     = "StringEquals"
+      variable = "autoscaling:ResourceTag/k8s.io/cluster-autoscaler/enabled"
+      values = [
+        "true"
+      ]
+    }
+  }
+}
+
+resource "helm_release" "cluster-autoscaler" {
+  depends_on = [
+    module.eks
+  ]
+
+  name       = "cluster-autoscaler"
+  namespace  = local.autoscaler_service_account_namespace
+  repository = "https://kubernetes.github.io/autoscaler"
+  chart      = "cluster-autoscaler"
+  version    = local.autoscaler_version
+
+  values = [yamlencode({
+    awsRegion = var.region
+    rbac = {
+      create = true
+      serviceAccount = {
+        name = local.autoscaler_service_account_name
+        annotations = {
+          "eks.amazonaws.com/role-arn" : module.autoscaler_iam_role.iam_role_arn
+        }
+      }
+    }
+    autoDiscovery = {
```

Comment:
Shouldn't we define `autoscalingGroups` instead of `autoDiscovery` to force limit for scaling?

Reply:
Why do you think this is a problem? I did follow the official documentation which is using autodiscovery:

https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/aws/README.md#auto-discovery-setup

> Auto-Discovery Setup is the preferred method to configure Cluster Autoscaler.

I am happy to change if it will be beneficial but I am not sure if we gain some improvement. The current setup also adheres to the maximum capacity of the used ASG. I have tested it by adding another node. Note the `3->4 (max: 10)` message:

```
I0308 06:19:48.682635       1 scale_up.go:468] Best option to resize: eks-appNode-m5_2xlarge-2abfb126-211a-9c36-aa0d-a9c663e33a07
I0308 06:19:48.682648       1 scale_up.go:472] Estimated 1 nodes needed in eks-appNode-m5_2xlarge-2abfb126-211a-9c36-aa0d-a9c663e33a07
I0308 06:19:48.682668       1 scale_up.go:595] Final scale-up plan: [{eks-appNode-m5_2xlarge-2abfb126-211a-9c36-aa0d-a9c663e33a07 3->4 (max: 10)}]
I0308 06:19:48.682677       1 scale_up.go:691] Scale-up: setting group eks-appNode-m5_2xlarge-2abfb126-211a-9c36-aa0d-a9c663e33a07 size to 4
```

Reply:
make sense